Fixed terrain tile culling problems when under ellipsoid

IanLilleyT opened the pull request:
Terrain tiles are culled when the horizon blocks them from view, as explained in [blog part 1](https://cesium.com/blog/2013/04/25/horizon-culling/) and [blog part 2](https://cesium.com/blog/2013/05/09/computing-the-horizon-occlusion-point/). But there's a problem when tiles are _under_ the ellipsoid. The second blog post says: 

> Also, please keep in mind an important caveat when using this approach. In the real world, objects occluded by the WGS84 ellipsoid are not necessarily occluded by the real surface of the Earth. This is because the Earth’s surface is actually slightly below the ellipsoid in parts of the world. Depending on your application, using WGS84 as the occluding volume may be acceptable, or you may need to use a more conservative ellipsoid.

Unfortunately, there are downsides to a more conservative ellipsoid: It increases the number of false positive draw calls and not all terrain providers have the up-front information of global min/max heights.

Instead, this PR does conservative ellipsoids per-tile. So if the tile's minimum height is below 0, it recomputes the tile's horizon occlusion point relative to a shrunk ellipsoid. This shrunk ellipsoid is shrunk just enough to be below the tile's lowest point, and no lower. During the tile culling pass it tests the tile's shrunk horizon occlusion point relative to its shrunk ellipsoid.

As shown in the picture below, the original horizon occlusion point is considered to be blocked by the horizon under existing code. In this PR, the new horizon occlusion point is relative to the shrunk ellipsoid and so is considered visible. (Side note: the shrunk horizon occlusion point would be in a slightly different position than the original. I forgot to draw it.)

![IMG_20191114_140300](https://user-images.githubusercontent.com/1328450/68895330-0a6fa880-06f7-11ea-8264-81fa9f3fc41f.jpg)

[And here's the fix in action in Death Valley](http://localhost:8080/Apps/Sandcastle/index.html#c=pVTLbtswEPwVwhcpgEy9ZSlxgra5tGiDBomRHqoeGHFdE6VJg6TkJEVu/bP+WKlX6kd6Ki8Cd3d2h8OhGqLQVipOF6AUYQKdo0vQrF7jSgEx8GUn556claIUjYU0arHqc9dKNoyCskAB2xF8dzPmB+xY5v4sBbKrVhydImdlzObU97fbLW7U1Kyg44IrufZ39r5hHHSIAxz4s9h3vL6HZUiZaduM5CkxBFWyVgb0I7pboKvfvz46pXjeYc5ge8S2i7lO1W0vpTC2GSjHQwNbs3+I0z3JvJf+vv+W0qEp+iD0BiojlZ3b9cfwYEBQdxjaB/vNS+0Ve2BiINuDdAUC8FJ+xyDIPQdqqS8J19AV3RJBK6INB0woXUjJ74m6AlG7X1+oP3QS9YMObgNNR7bdXaEhPQoshQZueVn8shaVYdJ6YNSkXQPHA3ksw119zvr6ZyvTjqA9q3/65P8pHFt0h8g3DzkDpFXL6SU/EL0ia1AEazCtQUbnUtCGCdIysaR2XHRJWtsxImJ3GsVRmmc5zvM8C+MiiT00TdI8yqMQx3FQpHGYRR6KZ1kyi3KcFnGWFfksik/GcysG1obHU94DoUx8v2amWt1Izt3ADkniIo7DPChmSRrO7KwAR2EbDaIkso2Tosg8ZJ9PEKVJUIRROkuCNA2LPD0Z3DvxJnNtHjlc9ATesPVGKtM+VBdj38B6w+3/QPv3dfUDDK60bmFzfwTNKWsQo+fl5OAZlRNUcaK1zSxrzm/ZE5STi7lv6/dgXHYn+9yA4uSxLVmFF5/6IMZ47tvtMcr0nrfle9G/d/vqrCcp1+9qY6y9Xs1rYo8Lr1QM3z8):

Before:
<img width="774" alt="Screen Shot 2019-11-14 at 3 46 20 PM" src="https://user-images.githubusercontent.com/1328450/68895064-992ff580-06f6-11ea-9fe2-4911ca5c8b6d.png">
After:
<img width="721" alt="Screen Shot 2019-11-14 at 3 46 27 PM" src="https://user-images.githubusercontent.com/1328450/68895070-9c2ae600-06f6-11ea-9ea4-1ea2217644d1.png">

Before:
<img width="789" alt="Screen Shot 2019-11-14 at 3 48 43 PM" src="https://user-images.githubusercontent.com/1328450/68896659-f24d5880-06f9-11ea-97e5-abb61c086a28.png">

After:
<img width="785" alt="Screen Shot 2019-11-14 at 3 48 35 PM" src="https://user-images.githubusercontent.com/1328450/68896670-f7aaa300-06f9-11ea-8ddf-8f5985cf9055.png">

A much better test case is if you manually change the `heightOffset` of a heightmap terrain provider to be -10000.0. This makes all tiles go under the ellipsoid and the culling is seriously broken. These problems are fixed in this PR and the correct number of tiles are rendered when checking the cesium inspector.

I'm happy to hear thoughts on this approach and the implementation details. All occurrences of `computeHorizonCullingPoint` have been replaced with `computeHorizonCullingPointPossiblyUnderEllipsoid` and `GlobeSurfaceTileProvider` now calls `isScaledSpacePointVisiblePossiblyUnderEllipsoid`. An alternative could be to do `if (minimumHeight < 0.0) { computeHorizonCullingPointUnderEllipsoid } else { computeHorizonCullingPoint }`, but this would happen in several areas so I thought it would be better to make the `Possibly` version. Also, I considered each tile storing its own shrunk `EllipsoidalOccluder`, but thought it wouldn't be worth the memory overhead. However, this would avoid the kind-of-funky changes to `EllipsoidalOccluder`. Besides the API changes there's also a small runtime cost with this new approach: each tile that is under the ellipsoid has to recalculate the camera position in scaled space every frame. It's not especially heavy math so I'm not too concerned.

So let me know what you think! I'll add tests once the design is more finalized.

cesium-concierge commented:
Thanks for the pull request @IanLilleyT!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/underEllipsoidCullingFix_TerrainTiles/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

IanLilleyT commented:
@kring if you have any thoughts

kring commented:
This looks like a good change to me, @IanLilleyT! It should be safe, in the sense that it can only result in fewer tiles being culled. And the ellipsoid adjustment should be small for normal terrains (highly exaggerated ones I'm less sure about), so it shouldn't lead to noticeably more tiles being rendered, either.

I guess you probably already realized this, but it's still imperfect even with this change, though. It's still possible for tiles to be improperly culled. Here's a sketch to illustrate the problem:

![image](https://user-images.githubusercontent.com/924374/68991401-8e21c480-08b2-11ea-917d-fd4f62455cf5.png)

Tile A is entirely above the ellipsoid, so no ellipsoid adjustment will happen. But tiles B, C, and D are below the ellipsoid. The camera is situated such that the ellipsoid is deemed to block the camera's view of tile A, so it gets culled. But in reality, A should be visible. This situation is admittedly a bit contrived, though, and this PR doesn't do anything to make it worse. I also think it'd be quite challenging to address this. It would almost certainly need to be done as part of the terrain processing rather than as a runtime tweak like in this PR.

So :shipit:!

IanLilleyT commented:
Hmm, maybe a hybrid approach. Find the min rendered terrain tile, like https://github.com/AnalyticalGraphicsInc/cesium/pull/8398, and use that. Would have to see what performance is like but it would definitely solve these edge case culling problems. Could even get a speedup when all tiles are above the earth, and use that larger ellipsoid to horizon cull more aggressively. Haven't fully thought it through though.

IanLilleyT commented:
Some more thoughts on this

The only way to horizon cull 100% correctly is to test each tile against the min tile's ellipsoid (see https://github.com/AnalyticalGraphicsInc/cesium/pull/8397#issuecomment-554622610). There a few tricky parts about doing this:
- The minimum rendered tile is not known until all tiles have been visited by the quadtree traversal, but the visibility check is part of the quadtree traversal, sooo.. you could use the previous frame's min tile, but what happens when a new tile enters view that is lower than last frame's min? Maybe it could use its own ellipsoid in that case.
- Need some way to recompute each tile's occlusion point relative to the smallest ellipsoid, cheaply. The [code](https://github.com/AnalyticalGraphicsInc/cesium/blob/098bbe7ae87d7f3b98389799640dfe32090fe1b0/Source/Core/EllipsoidalOccluder.js#L252) for generating the occlusion point isn't good to do per-tile, per-frame. Maybe we can precompute some of the values at tile-load-time that can get us partly there, such as `cosAlpha`, `sinAlpha`, and `magnitude`, and then be very careful with the math. Even with these optimizations I'm guessing [isScaledSpacePointVisible](https://github.com/AnalyticalGraphicsInc/cesium/blob/098bbe7ae87d7f3b98389799640dfe32090fe1b0/Source/Core/EllipsoidalOccluder.js#L118) will get at least two times heavier.

IanLilleyT commented:
@lilleyse ready for review

I refactored the `EllipsoidalOccluder` code and added some tests.

I also messed around for a while trying to trigger the pathological culling case, but couldn't get it. You would need a distant tile with a very abrupt height change. But the more distant a tile, the more likely it has some lower heights in it. So for now I'm moving forward with the original approach.

lilleyse commented on Source/Core/EllipsoidalOccluder.js:
```diff
@@ -116,16 +117,36 @@ import Rectangle from './Rectangle.js';
+                scratchCameraPositionInScaledSpaceShrunk
+            );
+
+            var distanceToLimbInScaledSpaceSquaredShrunk = Cartesian3.magnitudeSquared(cameraPositionInScaledSpaceShrunk) - 1.0;
```
This looks mostly similar to the code in the `cameraPosition` setter, aside from the `minimumHeight`. Can a shared function be used instead?

IanLilleyT replied:
It's unrolled for performance since it's called once per tile per frame (at least for tiles that are partly under ellipsoid). I added a comment to clarify.

IanLilleyT commented:
@lilleyse ready for another review. I was going to change the name to `isScaledSpaceTilePointVisible`
and `computeTileHorizonCullingPoint` instead of `isScaledSpacePointVisiblePossiblyUnderEllipsoid` and `computeHorizonCullingPointPossiblyUnderEllipsoid`, but I thought it would be misleading since nothing is forcing these functions to be used with tiles. It's a private function so we can do another PR if a better name comes up.

lilleyse commented:
Tested a few other sites, this seems to have fixed most of the culling bugs. Thanks @IanLilleyT
